### PR TITLE
[ENG-8039] Lakeview last modified approach for non blocking processing of commits

### PR DESCRIPTION
**Pull request by karankm97**
## What is the purpose of the pull request
Lakeview last modified approach for non blocking processing of commits

## Brief change log
  - *Added BLOCKING and NON_BLOCKING execution modes for active timeline*
  - *Not blocking execution on incomplete commits*
  - *Add new test cases*

## Verify this pull request
  - *Change covered by existing unit tests and new test cases*
  - *Testing to be done using real data*

### CODE REVIEW

**Comment by nimahajan** [comment]
Task linked: [ENG-8039 Metadata extractor - using last modified time based approach](https://app.clickup.com/t/18029943/ENG-8039)

File: src/main/java/com/onehouse/config/models/configv1/MetadataExtractorConfig.java [hunk 1]
```diff
@@ -23,6 +23,7 @@ public class MetadataExtractorConfig {
   @NonNull private List<ParserConfig> parserConfig;
   @Builder.Default private Optional<List<String>> pathExclusionPatterns = Optional.empty();
   @Builder.Default private JobRunMode jobRunMode = JobRunMode.CONTINUOUS;
+  @Builder.Default private UploadStrategy uploadStrategy = UploadStrategy.NON_BLOCKING;
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/config/models/configv1/MetadataExtractorConfig.java:26]: let default behaviour be blocking

File: src/main/java/com/onehouse/config/models/configv1/MetadataExtractorConfig.java [hunk 2]
```diff
@@ -47,4 +48,9 @@ public enum JobRunMode {
     CONTINUOUS,
     ONCE
   }
+
+  public enum UploadStrategy {
+    BLOCKING,
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/config/models/configv1/MetadataExtractorConfig.java:53]: nit: maybe we can rename this to BLOCK_ON_INCOMPLETE_COMMIT or something similar to be more clear about what we are blocking on?

File: src/main/java/com/onehouse/metadata_extractor/ActiveTimelineInstantBatcher.java [hunk 3]
```diff
@@ -107,9 +135,23 @@ public List<List<File>> createBatches(List<File> instants, int maxBatchSize) {
           currentBatch.clear();
           currentBatch.addAll(sortedInstants.subList(index, index + groupSize));
         }
-      } else {
-        // Instants are not related; add what we have and stop processing
-        shouldStopIteration = true;
+      } else if (!shouldStopIteration) {
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/metadata_extractor/ActiveTimelineInstantBatcher.java:138]: we should do this even if shouldStopIteration is true right? as shouldStopIteration is true when there are commits which are incomplete

File: src/main/java/com/onehouse/metadata_extractor/ActiveTimelineInstantBatcher.java [hunk 4]
```diff
@@ -128,30 +170,64 @@ public List<List<File>> createBatches(List<File> instants, int maxBatchSize) {
...
-    return batches;
+    return Pair.of(checkpoint, batches);
+  }
+
+  private static String decrementNumericString(String numericString) {
+    // Convert the string to a BigInteger
+    BigInteger number = new BigInteger(numericString);
+
+    // Decrement the value by 1
+    BigInteger decrementedNumber = number.subtract(BigInteger.ONE);
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/metadata_extractor/ActiveTimelineInstantBatcher.java:181]: the timestamps in the instant files are not guranteed to be 1ms appart. decrementing by 1 is not reliable

File: src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java [hunk 5]
```diff
@@ -251,9 +270,13 @@ private CompletableFuture<Checkpoint> uploadInstantsInSequentialBatches(
           Lists.partition(
               filesToUpload, getUploadBatchSize(CommitTimelineType.COMMIT_TIMELINE_TYPE_ARCHIVED));
     } else {
-      batches =
+      Pair<Checkpoint, List<List<File>>> checkpointBatchesPair =
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java:273]: If I am understanding correctly, on encountering an incomplete commit we dont process the subsequent commits in the current run but in the next extractor run it gets processed?

File: src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java [hunk 6]
```diff
@@ -251,9 +270,13 @@ private CompletableFuture<Checkpoint> uploadInstantsInSequentialBatches(
...
           activeTimelineInstantBatcher.createBatches(
-              filesToUpload, getUploadBatchSize(CommitTimelineType.COMMIT_TIMELINE_TYPE_ACTIVE));
+              filesToUpload,
+              getUploadBatchSize(CommitTimelineType.COMMIT_TIMELINE_TYPE_ACTIVE),
+              checkpoint);
+      batches = checkpointBatchesPair.getRight();
+      checkpoint = checkpointBatchesPair.getLeft();
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java:279]: can we return just the last incompleteCommitInstant file name? here we are updating the checkpoint obj and in later parts of the code we refer to it as the previous checkpoint which that is not correct (as it has been modified)

File: src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java [hunk 7]
```diff
@@ -127,8 +128,22 @@ public CompletableFuture<Checkpoint> paginatedBatchUploadWithCheckpoint(
             storageUtils.constructFileUri(
                 table.getAbsoluteTableUri(), getPathSuffixForTimeline(commitTimelineType)));
 
+    // Compute the start after for the entire paginated batched upload. This handles the case where
```
Comment (sampan-s-nayak) [inline src/main/java/com/onehouse/metadata_extractor/TimelineCommitInstantsUploader.java:131]: I didnt understand this

**Review by sampan-s-nayak — APPROVED** [review]

**Review by karankm97 — COMMENTED** [review]
/push-image lakeview-nonblock-approach

**Review by karankm97 — COMMENTED** [review]
/push-image lakeview-nonblock-approach

**Pull request merged**